### PR TITLE
introducing srpm

### PR DESCRIPTION
**Pull request by dparalen**
Doesn't work yet :-/

### CODE REVIEW

File: pulpsolv.py [hunk 1]
```diff
@@ -327,16 +327,39 @@ def __call__(self, unit):
         return solvable
 
 
+class SrpmSolvableFactory(BasetUnitSolvableFactory):
+    # An SRPM content unit factory
+    attribute_factories = [
+        AttributeFactory('name'),
+        EVRAttributeFactory(),
+        # FIXME: any other arch makes sense?
+        AttributeFactory('arch', conversion=lambda x: 'noarch'),
```
Comment (dparalen) [inline pulpsolv.py]: FIXME**2 works the moment I set arch to noarch... I don't mind but I sorta hope there's a better way O:-)

**Comment by dparalen** [comment]
Got it working somehow, had to trick the `src` to `source` arch and  call `pool.setarch('source')` for the solver to be able to find the package.
There's still the issue of a srpm having build requirements i.e `source != x86_64`; would that solve?

**Comment by dparalen** [comment]
Just filed [an issue with `libsolv`](https://github.com/openSUSE/libsolv/issues/269) to figure out the right approach; leaving this open till I get some feedback I guess is the best right now.

**Comment by igor-raits** [comment]
let me look at this.

File: pulpsolv.py [hunk 2]
```diff
@@ -327,16 +327,38 @@ def __call__(self, unit):
         return solvable
 
 
+class SrpmSolvableFactory(BasetUnitSolvableFactory):
+    # An SRPM content unit factory
+    attribute_factories = [
+        AttributeFactory('name'),
+        EVRAttributeFactory(),
+        # FIXME: the default src doesn't work for me
+        AttributeFactory('arch', conversion=lambda x: 'source'),
```
Comment (igor-raits) [inline pulpsolv.py]: it should be `src`, fullstop

**Review by igor-raits — COMMENTED** [review]
it doesn't find `src` solvables because those are not installable. you should do something like selection, pass `SELECTION_WITH_SOURCE` and that's it... and probably with SELECTION_DOTARCH

**Comment by dparalen** [comment]
Hey @ignatenkobrain!

Thanks for the tips!
/me updates PR...

Cheers,
milan

**Comment by mlschroe** [comment]
Note that you *can* address src packages (i.e. packages that have arch="src" or arch="nosrc") by creating a .arch dependency:
```
id = pool.str2id(name, create=True)
id = pool.rel2id(id, solv.ARCH_SRC, solv.REL_ARCH, create=True)
job = pool.Job(solv.Job.SOLVER_SOLVABLE_NAME | solv.Job.SOLVER_INSTALL, id)
```

**Comment by dparalen** [comment]
@mlschroe I'm trying to wrap my head around your suggestion, is it like I suppose to set this for the job request or every time I create a srpm solvable?
My guess is the former...

Thanks!

**Comment by mlschroe** [comment]
Yes, it is for the job request. A job selects a set of packages, usually src/nosrc packages are ignored. With that REL_ARCH dependency you're limiting the packages to the specified architecture. There's a small hack in libsolv that makes this work for src/nosrc packages as well (they are excluded by default).

**Comment by dparalen** [comment]
@mlschroe I see, thanks!
I've just tried that and it works indeed.
There's a limitation though I'd have to guess what arch is the solvable, the user is requesting to figure out the dependencies of, so in this POC I guess I'm better with leaving the guessing on libsolv thru the selection&flags (which I didn't know before how to use).
Outside of this POC I'll have that information though, so I guess I'll go that way to limit the guessing.

Thanks a ton for the input, @ignatenkobrain, @mlschroe! 

--
milan

**Comment by dparalen** [comment]
I guess this is good to go merge...

**Pull request merged**